src/coreclr/jit/morph.cpp

```diff
@@ -9922,6 +9926,7 @@ GenTree* Compiler::fgOptimizeCast(GenTreeCast* cast)
 
             return src;
         }
+#endif
```

Comment:
This is almost certainly not the right fix. Transformations made in morph generally rely on IR invariants that must be the same regardless of which backend is eventually used to emit the code. The fact that you have to comment out transformations in morph should usually be an indication that there is a bug in the RISCV64 backend. Can you please share the problematic IR change that this transformation is making so that I can confirm whether this is a legal transformation or not on JIT IR?

Reply:
I made this change in the way I did in order to not change behaviour in the other ports,

If, as you say, "transformations made in morph rely on IR invariants that must be the same regardless of which backend is eventually used to emit the code" then the only conclusion I can come to is that this machine-dependent tranformation of optimising out narrowing casts should not be here in the first place, and the correct fix is to delete it from here, and implement it in code that allows machine-dependent transformations, perhaps the individual backends.

Narrowing casts are not, in an abstract machine-independent sense, NOPs because of at least:

- in a checked context, overflow must be checked for (this also applies to same-size casts between signed and unsigned in either direction, and to widening casts from signed to unsigned). The code already bails out in this case

- the narrowed value requires at least sign-extension or zero-extension to int/uint before use by future operations. This is required by the C#/CIL spec. RISC-V additionally requires sign extension (always) to 64 bits

- both the RISC-V instruction set and the ABI are designed around values being made canonical (to 64 bits) immediately, so that you can as one example pass them in a 64 bit register as a function argument or result and the receipient can use them immediately in e.g. a compare-and-branch. The x86 and Arm strategy seems to be to extract and sign/extend the value on use.

- it's possible to imagine some machine in which a narrowing cast requires moving the value to a different register

In short: removing a narrowing cast is not a machine-independent transformation. (and nor is removing a widening cast)


> an indication that there is a bug in the RISCV64 backend

If the cast is removed then the backend never sees it, so can not be at fault. If the cast is allowed to remain then the backend generates the appropriate sign/zero-extension instructions.

> Can you please share the problematic IR change

As above, a minimised example that triggers the bad code generation is:

    using System.Runtime.CompilerServices;
    
    public static object bug(IRuntimeVariables vars)
    {
        return vars[0];
    }

This interacts with the code in the System.Runtime.CompilerServices.RuntimeOps.RuntimeVariableList.GetStrongBox` method (RuntimeOps.RuntimeVariableList.cs:97):

    private IStrongBox GetStrongBox(int index)
    {
        // We lookup the closure using two ints:
        // 1. The high dword is the number of parents to go up
        // 2. The low dword is the index into that array
        long closureKey = _indexes[index];
    
        // walk up the parent chain to find the real environment
        object[] result = _data;
        for (int parents = (int)(closureKey >> 32); parents > 0; parents--)
        {
            result = HoistedLocals.GetParent(result);
        }
    
        // Return the variable storage
        return (IStrongBox)result[unchecked((int)closureKey)];
    }

The `closureKey` is a 64 bit value that actually contains two 32 bit values. By far the most common case here is that the closure has only a single scope and the upper 32 bits are zero. But the failing MixedScope test explicitly creates a closure with nested scopes, which means the upper 32 bits are non-zero for some values of `index`.

The problem arises in:

    result[unchecked((int)closureKey)]

The cast to `int` is intended to discard the upper bits so that when the result is extended to 64 bits (and shifted) to index into `result`, only the low bits of `closureKey` will be used.

According to the semantics of C# this code is correct. The `unchecked` is unnecessary as that is that default for non-const arithmetic, but perhaps serves as a reminder that a checked cast would be expected to fail here because the upper bits can be non-zero.

The code explicitly casts a 64 bit value to 32 bits, and then the array indexing implicitly casts it back to 64 bits.

Optimising out the cast to 32 bits is incorrect on RISC-V in particular, because then no one sign-extends the lower 32 bits, but it is also incorrect in the abstract.

It also appears to be incorrect on loongarch. If this incorrect optimisation is removed from morph then commit 7897367882b76228cfbba1d0d9c13030b5bedc99 will also not be necessary (it's just the same work-around as was previously applied for RISC-V and which I'm removing in this PR).

The `return` basic block starts as IL:

    IL_0027  07                ldloc.1
    IL_0028  06                ldloc.0
    IL_0029  69                conv.i4
    IL_002a  9a                ldelem.ref
    IL_002b  74 11 00 00 01    castclass    0x1000011
    IL_0030  2a                ret

The morph processing is:

    fgMorphTree BB04, STMT00007 (before)
                   [000035] --CXG------                         *  RETURN    ref
                   [000034] --CXG------                         \--*  CALL help ref    CORINFO_HELP_CHKCASTINTERFACE
                   [000033] H------N--- arg0                       +--*  CNS_INT(h) long   0x3f2b662b28 class
                   [000032] n--XG------ arg1                       \--*  IND       ref
                   [000031] ---XG------                               \--*  INDEX_ADDR byref ref[]
                   [000028] -----------                                  +--*  LCL_VAR   ref    V03 loc1          (last use)
                   [000030] -----------                                  \--*  CAST      int <- long
                   [000029] -----------                                     \--*  LCL_VAR   long   V02 loc0          (last use)
    Initializing arg info for 34.CALL:
    Args for call [000034] CALL after AddFinalArgsAndDetermineABIInfo:
    CallArg[[000033].CNS_INT long (By value), 1 reg: a0, byteAlignment=8]
    CallArg[[000032].IND ref (By value), 1 reg: a1, byteAlignment=8]
    
    Morphing args for 34.CALL:
    BB04 requires throw helper block for SCK_RNGCHK_FAIL
    fgMorphIndexAddr (before remorph):
                   [000111] ---X-O-----                         *  COMMA     byref
                   [000103] ---X-O-----                         +--*  BOUNDS_CHECK_Rng void
                   [000030] -----------                         |  +--*  CAST      int <- long
                   [000029] -----------                         |  |  \--*  LCL_VAR   long   V02 loc0          (last use)
                   [000102] ---X-------                         |  \--*  ARR_LENGTH int
                   [000028] -----------                         |     \--*  LCL_VAR   ref    V03 loc1          (last use)
                   [000110] -----O-----                         \--*  ARR_ADDR  byref ref[]
                   [000109] -----------                            \--*  ADD       byref
                   [000099] -----------                               +--*  LCL_VAR   ref    V03 loc1          (last use)
                   [000108] -----------                               \--*  ADD       long
                   [000106] -----------                                  +--*  MUL       long
                   [000104] ---------U-                                  |  +--*  CAST      long <- uint
                   [000100] -----------                                  |  |  \--*  CAST      int <- long
                   [000101] -----------                                  |  |     \--*  LCL_VAR   long   V02 loc0          (last use)
                   [000105] -------N---                                  |  \--*  CNS_INT   long   8
                   [000107] -----------                                  \--*  CNS_INT   long   16
    GenTreeNode creates assertion:
                   [000102] ---X-+-----                         *  ARR_LENGTH int
    In BB04 New Local Constant Assertion: V03 != null, index = #01
    BB04 requires throw helper block for SCK_RNGCHK_FAIL
    fgMorphIndexAddr (after remorph):
                   [000111] ---X-+-----                         *  COMMA     byref
                   [000103] ---X-+-----                         +--*  BOUNDS_CHECK_Rng void
                   [000029] -----+-----                         |  +--*  LCL_VAR   int    V02 loc0          (last use)
                   [000102] ---X-+-----                         |  \--*  ARR_LENGTH int
                   [000028] -----+-----                         |     \--*  LCL_VAR   ref    V03 loc1          (last use)
                   [000110] -----+-----                         \--*  ARR_ADDR  byref ref[]
                   [000109] -----+-----                            \--*  ADD       byref
                   [000099] -----+-----                               +--*  LCL_VAR   ref    V03 loc1          (last use)
                   [000108] -----+-----                               \--*  ADD       long
                   [000106] -----+-----                                  +--*  LSH       long
                   [000104] -----+---U-                                  |  +--*  CAST      long <- uint
                   [000101] -----+-----                                  |  |  \--*  LCL_VAR   int    V02 loc0          (last use)
                   [000105] -----+-N---                                  |  \--*  CNS_INT   long   3
                   [000107] -----+-----                                  \--*  CNS_INT   long   16

The local variable V02, which is in a register,  was known to be a long and then was cast to int, which means it will be sign-extended. After the last transformation it is simply assumed to be an int, which is incorrect as bits 63:31 are not all the same in certain cases -- that is the reason for the cast to int.

If the variable was in memory and loaded with a `lw` instruction then that would sign-extend it. But as it is already in a register, simply pretending it is an `int` doesn't work. Note that the "long <- uint" cast is, by design, a NOP on RISC-V. Both `int` and `uint` are expected to be already sign-extended to 64 bits when in a register. A "ulong <- uint" cast requires zero-extension.

Note: the existence of the "long <- uint" cast instead of a "long <- int" cast is also wrong, but fortunately harmless as C# requires array indexes to be less than 2^31. It is claimed in the code that does it to be "more efficient", but that is a machine-dependent assumption.

Reply:
> Optimising out the cast to 32 bits is incorrect on RISC-V in particular, because then no one sign-extends the lower 32 bits, but it is also incorrect in the abstract.
> 
> It also appears to be incorrect on loongarch. If this incorrect optimisation is removed from morph then commit [7897367](https://github.com/dotnet/runtime/commit/7897367882b76228cfbba1d0d9c13030b5bedc99) will also not be necessary (it's just the same work-around as was previously applied for RISC-V and which I'm removing in this PR).
> 
> The local variable V02, which is in a register, was known to be a long and then was cast to int, which means it will be sign-extended. After the last transformation it is simply assumed to be an int, which is incorrect as bits 63:31 are not all the same in certain cases -- that is the reason for the cast to int.
> 
> If the variable was in memory and loaded with a `lw` instruction then that would sign-extend it. But as it is already in a register, simply pretending it is an `int` doesn't work. Note that the "long <- uint" cast is, by design, a NOP on RISC-V. Both `int` and `uint` are expected to be already sign-extended to 64 bits when in a register. A "ulong <- uint" cast requires zero-extension.
> 

The semantics of the IR shown here requires the backend to access only the lower 32 bits of `V02`. Under those semantics the truncation is a noop and thus the transformation is legal. If RISCV64 accesses the upper bits then it is a bug in the RISCV64 backend. Disabling the transformation for RISC-V here is not a fix for the problem; other places in the JIT can also legally make use of this assumption, and the frontend and middle-end should not be involved in accounting for semantics mismatches in the backends.

`TYP_INT` access of `TYP_LONG` locals is allowed in RyuJIT IR. It is motivated by the fact that our primary backends can efficiently address the lower 32 bits of a 64-bit register. If this is not the case on RISC-V then the RISC-V backend should be the place to make the necessary adjustments. Concretely it could mean that you insert the extensions again during RISCV64 specific lowering when you see a `TYP_INT` use of a `TYP_LONG` local.

The LA64 code looks appropriate to me assuming their backend has the same semantics as our x64 and arm64 backends, which is that the register assigned to a `TYP_INT` node has undefined upper bits. Again, if you provide different guarantees in the RISCV64 backend then that's the place you need to make sure you are providing those guarantees, not in the front end/middle end.

> Note: the existence of the "long <- uint" cast instead of a "long <- int" cast is also wrong, but fortunately harmless as C# requires array indexes to be less than 2^31. It is claimed in the code that does it to be "more efficient", but that is a machine-dependent assumption.

This is a legal transformation as bounds checks have already ensured that the last bit is zero.

Reply:
Well, the choices of the IR semantics are historical and motivated by IL and by our supported targets. The IR only supports integral operations on 32-bit and 64-bit values which maps well from IL and maps well into our primary targets. Casts are required to explicitly represent the smaller truncations or type changes; no IR node can produce a value of a small type, those are always considered to produce at least a 32 bit value. The `TYP_INT` access of `TYP_LONG` locals is allowed as a special case and the motivation is what I mention above (our primary targets directly support 32-bit operations that don't touch the upper bits).

It would be perfectly possible to change the semantics so that `TYP_INT` access of `TYP_LONG` locals is disallowed (i.e. move this piece of morph into the x64/arm64 backends instead). We would need to find the places that make use of it and fix them (to be fair there probably aren't that many, but e.g. #97865 is adding a new one). It would also require agreement from the team that we should do that and would require work to make sure that all such patterns get handled as efficiently in x64 and arm64 as they are today. I would not personally be opposed to this change in the IR semantics assuming it doesn't regress things, but it would be a discussion that is separate from this PR.

Reply:
FWIW, I view this `TYP_INT`/`TYP_LONG` type mismatch on locals as an old x86 "quick CQ hack". There is only one place that produces it right now - `optNarrowTree`, and it would be perfectly possible to represent this with a narrowing cast everywhere.